{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# align isoforms to either update genome or annotation file for full length BHF\n",
    "# Look for inverted repeat regions (see ppt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {},
   "outputs": [],
   "source": [
    "import genomeview\n",
    "import pyBigWig\n",
    "import subprocess, os, glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from IPython.display import display, HTML\n",
    "\n",
    "def create_pwm(cov_path, min_cov = None, fract_mean_cov = 0.10):\n",
    "    \"\"\"returns PWM\"\"\"\n",
    "    rev_comp = {'a':'T', 't':'A', 'c':'G', 'g':'C'}\n",
    "    qual_dict = {}\n",
    "    for idx, x in enumerate(range(300)): # ASCII -33 (PHRED)\n",
    "        qual_dict[chr(x)] = idx-33\n",
    "\n",
    "    chrom_list = []\n",
    "    pos_list = []\n",
    "    ref_list = []\n",
    "    baseq_list = []\n",
    "    mapq_list = []\n",
    "    A_list = []\n",
    "    T_list = []\n",
    "    G_list = []\n",
    "    C_list = []\n",
    "    cbase_list = []\n",
    "    \n",
    "    reads_list = []\n",
    "    for line in open(cov_path):\n",
    "        ID,pos,ref,num_reads,bases,base_qual,map_qual = line.split()\n",
    "        reads_list = reads_list + [int(num_reads)]\n",
    "    mean_cov = np.mean(reads_list)\n",
    "    if min_cov == None:\n",
    "        cov_threshold = int(fract_mean_cov*mean_cov)\n",
    "    else: \n",
    "        cov_threshold = int(min_cov)\n",
    "    print('coverage_threshold = ', cov_threshold)\n",
    "\n",
    "    for line in open(cov_path):\n",
    "        ID,pos,ref,num_reads,bases,base_qual,map_qual = line.split()\n",
    "        chrom_list = chrom_list + [ID]\n",
    "        pos_list = pos_list + [pos]\n",
    "        ref_list = ref_list + [ref]\n",
    "        baseq_list = baseq_list + [int(np.median([qual_dict[x] for x in base_qual]))]\n",
    "        mapq_list = mapq_list + [int(np.median([qual_dict[x] for x in map_qual]))]\n",
    "\n",
    "        if int(num_reads) == 0: # at least 1 read coverage\n",
    "            A_list = A_list + [0]\n",
    "            T_list = T_list + [0]\n",
    "            G_list = G_list + [0]\n",
    "            C_list = C_list + [0]\n",
    "            cbase_list = cbase_list + [ref]\n",
    "        else:\n",
    "            counts = {'A':0, 'C':0, 'G':0, 'T':0}\n",
    "            for c in bases:\n",
    "                if c in {'.', ','}:\n",
    "                    counts[ref] += 1\n",
    "                elif c in {'A', 'C', 'G', 'T'}:\n",
    "                    counts[c] += 1\n",
    "                elif c in {'a', 'c', 'g', 't'}:\n",
    "                    counts[rev_comp[c]] += 1\n",
    "            if sum(counts.values()) == 0: # at least 1 base detected\n",
    "                A_list = A_list + [0]\n",
    "                T_list = T_list + [0]\n",
    "                G_list = G_list + [0]\n",
    "                C_list = C_list + [0]\n",
    "                cbase_list = cbase_list + [ref]\n",
    "            else:\n",
    "                A_list = A_list + [counts['A']]\n",
    "                T_list = T_list + [counts['T']]\n",
    "                G_list = G_list + [counts['G']]\n",
    "                C_list = C_list + [counts['C']]\n",
    "                cbase = [key for key,value in counts.items() if value >= cov_threshold]\n",
    "                if len(cbase) > 1:\n",
    "                    cbase = ['/'.join(cbase)]\n",
    "                elif len(cbase) == 0:\n",
    "                    cbase = [ref]\n",
    "                cbase_list = cbase_list + cbase\n",
    "\n",
    "    pwm = pd.DataFrame({'chrom':chrom_list,\n",
    "                        'pos':pos_list,\n",
    "                        'ref':ref_list,\n",
    "                        'baseq':baseq_list,\n",
    "                        'mapq':mapq_list,\n",
    "                        'chrom':chrom_list,\n",
    "                        'A':A_list,\n",
    "                        'T':T_list,\n",
    "                        'G':G_list,\n",
    "                        'C':C_list,\n",
    "                        'cbase':cbase_list\n",
    "                       })\n",
    "    pwm['match'] = [x==y for x,y in zip(pwm['ref'], pwm['cbase'])]\n",
    "    \n",
    "    return pwm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_15_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_16_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_17_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_1_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_22_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_24_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_25_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_26_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_2_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_31_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_34_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_35_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_38_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_3_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_47_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_48_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_49_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_55_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_56_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_58_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_59_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_66_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_67_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_6_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_72_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_73_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_75_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_76_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_81_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_82_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_86_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_87_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_89_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_91_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_93_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_94_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_95_Aligned.sortedByCoord.out.bam',\n",
       " '/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_97_Aligned.sortedByCoord.out.bam']"
      ]
     },
     "execution_count": 249,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aligner = 'STAR'\n",
    "\n",
    "if aligner == 'STAR':\n",
    "    output_dir = '/mnt/ibm_lg/daniel_le/data/botryllus/results/'\n",
    "elif aligner == 'BWA':\n",
    "    output_dir = '/mnt/ibm_lg/daniel_le/data/botryllus/coverage/bwa_coverage/'\n",
    "\n",
    "completed_targets = glob.glob(f'{output_dir}*.bam')\n",
    "sorted(completed_targets)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "create pileup...\n"
     ]
    }
   ],
   "source": [
    "sample_index = 48\n",
    "print_result = False\n",
    "\n",
    "fasta_path = '/mnt/ibm_lg/daniel_le/data/botryllus/genome/botznik-chr.fa'\n",
    "genebed_path = '/mnt/ibm_lg/daniel_le/data/botryllus/genome/botznik-chr-all.bed'\n",
    "if aligner == 'STAR':\n",
    "    bam_path = f'{output_dir}sample_{sample_index}_Aligned.sortedByCoord.out.bam'\n",
    "    bw_path = f'{output_dir}sample_{sample_index}_Aligned.sortedByCoord.out.bam.bw'\n",
    "elif aligner == 'BWA':\n",
    "    bam_path = f'{output_dir}/bwa_alignments/sample_{sample_index}.sorted.bam'\n",
    "    bw_path = f'{output_dir}sample_{sample_index}_bs1.bw'\n",
    "    \n",
    "dataset_paths = [\n",
    "                 bam_path,\n",
    "                 bw_path,\n",
    "                 genebed_path,\n",
    "                ]\n",
    "\n",
    "# BHF = chr9:7538434..7544266\n",
    "chrom = \"chr9\"\n",
    "start = 7538434\n",
    "end =   7544266\n",
    "\n",
    "# create pileup\n",
    "if aligner == 'BWA':\n",
    "    cov_dir = '/mnt/ibm_lg/daniel_le/data/botryllus/coverage/bwa_slice_cov/'\n",
    "elif aligner == 'BWA':\n",
    "    cov_dir = '/mnt/ibm_lg/daniel_le/data/botryllus/coverage/star_slice_cov/'\n",
    "cov_path = f'{cov_dir}sample_{sample_index}_{chrom}:{start}-{end}.pileup'\n",
    "    \n",
    "if not os.path.exists(cov_path):\n",
    "    print('create pileup...')\n",
    "    subprocess.call(['samtools',\n",
    "                     'mpileup',\n",
    "                     '-s',\n",
    "                     '-f',\n",
    "                     fasta_path,\n",
    "                     '-r',\n",
    "                     f'{chrom}:{start}-{end}',\n",
    "                     bam_path,\n",
    "                     '-o',\n",
    "                     cov_path\n",
    "                    ])\n",
    "\n",
    "if print_result == True:\n",
    "    doc = genomeview.visualize_data(dataset_paths, chrom, start, end, fasta_path)\n",
    "    display(doc)\n",
    "    \n",
    "    pwm = create_pwm(cov_path)\n",
    "    pwm[pwm['match'] == False]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bam processed\n",
      "Output: \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "n_threads = 16\n",
    "tmp_dir = '/mnt/ibm_lg/daniel_le/data/botryllus/tmp_splits/'\n",
    "output_splits_dir = '/mnt/ibm_lg/daniel_le/data/botryllus/star_splits/'\n",
    "\n",
    "target = f'/mnt/ibm_lg/daniel_le/data/botryllus/results/sample_{sample_index}_Aligned.sortedByCoord.out.bam'\n",
    "target_fn_prefix = target.split('/')[-1].split('.bam')[0]\n",
    "output_splitbam = f'{output_splits_dir}{target_fn_prefix}.splitBam.bam'\n",
    "output_splitbw = f'{output_splits_dir}{target_fn_prefix}.splitBam.bw'\n",
    "output_splitpileup = f'{output_splits_dir}{target_fn_prefix}.splitBam.pileup'\n",
    "\n",
    "# filter bam file for split reads\n",
    "p1 = subprocess.Popen(['samtools',\n",
    "                     'view',\n",
    "                       '-h',\n",
    "                     '-F', # only unique reads\n",
    "                     '256',\n",
    "                     target,\n",
    "                     f'{chrom}:{start}-{end}',\n",
    "                    ], shell=False, stdin=subprocess.PIPE, stdout=subprocess.PIPE)\n",
    "p2 = subprocess.Popen([f\"awk '$6 ~ /N/ && $4 >= {start} || $1 ~ /^@/'\", # only split reads thats start within range\n",
    "                      ], shell=True, stdin=p1.stdout, stdout=subprocess.PIPE)\n",
    "\n",
    "with open(output_splitbam, 'w') as outfile:\n",
    "    p3 = subprocess.Popen(['samtools',\n",
    "                             'view',\n",
    "                             '-S',\n",
    "                             '-b',\n",
    "                             '-',\n",
    "                            ], shell=False, stdin=p2.stdout, stdout=outfile)\n",
    "print('bam processed')\n",
    "import time\n",
    "time.sleep(1)\n",
    "\n",
    "# index bam\n",
    "cmnd = f'samtools index {output_splitbam}'\n",
    "try:\n",
    "    output = subprocess.check_output(\n",
    "        cmnd, stderr=subprocess.STDOUT, shell=True, timeout=3,\n",
    "        universal_newlines=True)\n",
    "except subprocess.CalledProcessError as exc:\n",
    "    print(\"Status : FAIL\", exc.returncode, exc.output)\n",
    "else:\n",
    "    print(\"Output: \\n{}\\n\".format(output))\n",
    "    \n",
    "# create bigwig track\n",
    "bin_size = 1\n",
    "subprocess.call(['bamCoverage',\n",
    "                '-p',\n",
    "                str(n_threads),\n",
    "                '--normalizeUsing',\n",
    "                'CPM',\n",
    "                '--ignoreDuplicates',\n",
    "                '--binSize',\n",
    "                str(bin_size),\n",
    "                '-b',\n",
    "                output_splitbam,\n",
    "                '-o',\n",
    "                output_splitbw,\n",
    "                ])\n",
    "print('bigwig created')\n",
    "\n",
    "# create pileup\n",
    "subprocess.call(['samtools',\n",
    "                 'mpileup',\n",
    "                 '-s',\n",
    "                 '-f',\n",
    "                 fasta_path,\n",
    "                 '-r',\n",
    "                 f'{chrom}:{start}-{end}',\n",
    "                 output_splitbam,\n",
    "                 '-o',\n",
    "                 output_splitpileup\n",
    "                ])\n",
    "print('pileup created')\n",
    "    \n",
    "# display coverage\n",
    "dataset_paths = [\n",
    "                 output_splitbam,\n",
    "                 output_splitbw,\n",
    "                 genebed_path,\n",
    "                ]\n",
    "\n",
    "doc = genomeview.visualize_data(dataset_paths, chrom, start, end, fasta_path)\n",
    "display(doc)\n",
    "\n",
    "pwm = create_pwm(output_splitpileup, min_cov = 1)\n",
    "pwm[[a!=0 or t!=0 or g!=0 or c!=0 for a,t,g,c in zip(pwm['A'],\n",
    "                                                        pwm['T'],\n",
    "                                                        pwm['G'],\n",
    "                                                        pwm['C'],)]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mypython3",
   "language": "python",
   "name": "mypython3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
