{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "import shutil, os, subprocess\n",
    "import pandas as pd\n",
    "\n",
    "def translate(seq, frame=1): \n",
    "    # frame: 1 = start at pos 0; 2 = start at pos 1; 3 = start at pos 2\n",
    "    table = { \n",
    "        'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M', \n",
    "        'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T', \n",
    "        'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K', \n",
    "        'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                  \n",
    "        'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L', \n",
    "        'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P', \n",
    "        'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q', \n",
    "        'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R', \n",
    "        'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V', \n",
    "        'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A', \n",
    "        'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E', \n",
    "        'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G', \n",
    "        'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S', \n",
    "        'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L', \n",
    "        'TAC':'Y', 'TAT':'Y', 'TAA':'', 'TAG':'', \n",
    "        'TGC':'C', 'TGT':'C', 'TGA':'', 'TGG':'W', \n",
    "    } \n",
    "    comp_dict = {'C':'G',\n",
    "                 'G':'C',\n",
    "                 'A':'T',\n",
    "                 'T':'A'\n",
    "                }\n",
    "    protein = ''\n",
    "    \n",
    "    if frame == 1 :\n",
    "        start_val = 0\n",
    "    if frame == 2:\n",
    "        start_val = 1\n",
    "    if frame == 3:\n",
    "        start_val = 2\n",
    "    if frame == 4 :\n",
    "        start_val = 0\n",
    "    if frame == 5:\n",
    "        start_val = 1\n",
    "    if frame == 6:\n",
    "        start_val = 2\n",
    "    if frame > 3:\n",
    "        seq = ''.join([comp_dict.get(x) for x in seq])\n",
    "        \n",
    "    for i in range(start_val, len(seq), 3): \n",
    "        try:\n",
    "            codon = seq[i:i + 3] \n",
    "            protein+= table[codon] \n",
    "        except:\n",
    "            break\n",
    "    return protein\n",
    "\n",
    "def split_translation (full_seq, len_cutoff = 10):\n",
    "    sub_seq = [x for x in full_seq.split('_') if len(x) > len_cutoff]\n",
    "    return sub_seq\n",
    "\n",
    "def print_fasta (sub_list, prefix='test'):\n",
    "    for idx, entry in enumerate(sub_list):\n",
    "        print(f'>{prefix}_{idx}\\n{entry}')\n",
    "        \n",
    "def save_fasta (sub_list, prefix, output):\n",
    "    with open(output, 'w') as f:\n",
    "        for idx, entry in enumerate(sub_list):\n",
    "            line = f'>{prefix}_{idx}\\n{entry}\\n'\n",
    "            f.write(line)\n",
    "            \n",
    "def process_fasta (infile, frames, outfile):\n",
    "    with open(infile, 'r') as f:\n",
    "        sublines = []\n",
    "        entry_name = ''\n",
    "        for idx,line in enumerate(f):\n",
    "            line = line[:-1]\n",
    "            if line.startswith('>') and entry_name == '':\n",
    "                entry_name = line\n",
    "            if line.startswith('>') and entry_name != '':\n",
    "                if len(sublines) > 0:\n",
    "                    print_seq = ''.join(sublines)\n",
    "                    ### process entry seq\n",
    "                    with open(outfile, 'a') as outf:\n",
    "                        for frame in frames:\n",
    "                            full_seq = translate(print_seq, frame)\n",
    "                            outf.write(f'{entry_name}__fr_{frame}\\n{full_seq}\\n')\n",
    "                    ###\n",
    "                    sublines = []\n",
    "                entry_name = line\n",
    "            else:\n",
    "                sublines = sublines+[line]                          \n",
    "        if len(sublines) > 0:\n",
    "            print_seq = ''.join(sublines)\n",
    "            ### process entry seq\n",
    "            with open(outfile, 'a') as outf:\n",
    "                for frame in frames:\n",
    "                    full_seq = translate(print_seq, frame)\n",
    "                    outf.write(f'{entry_name}__fr_{frame}\\n{full_seq}\\n')\n",
    "            ###\n",
    "            sublines = []\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "deleted existing results subdir\n",
      "created results subdir\n",
      "translated to protein\n",
      "blastp -query /home/ubuntu/data/longread/blast_results/A10_TSO1/A10_TSO1_protein.fa -db /home/ubuntu/data/longread/proteome/UP000005640_9606.fasta -outfmt \"6 qseqid sseqid evalue\" -max_target_seqs 1 -out /home/ubuntu/data/longread/blast_results/A10_TSO1/A10_TSO1_blastp_out.txt -num_threads 4\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_dir = '/home/ubuntu/data/longread/blast_results/'\n",
    "target_path = '/home/ubuntu/data/longread/A10_TSO1.fa'\n",
    "target_suffix = '.fa'\n",
    "cell_id = target_path.split('/')[-1].split(target_suffix)[0]\n",
    "frames = [1,2,3]\n",
    "proteome_ref = '/home/ubuntu/data/longread/proteome/UP000005640_9606.fasta'\n",
    "n_threads = 4\n",
    "\n",
    "# create results subdir\n",
    "try:\n",
    "    shutil.rmtree(f'{results_dir}{cell_id}')\n",
    "    print('deleted existing results subdir')\n",
    "except:\n",
    "    pass\n",
    "try:\n",
    "    os.mkdir(f'{results_dir}{cell_id}')\n",
    "    print('created results subdir')\n",
    "except:\n",
    "    pass\n",
    "\n",
    "# translate nucl to prot\n",
    "process_fasta(f'{target_path}', \n",
    "              frames,\n",
    "              f'{results_dir}{cell_id}/{cell_id}_protein.fa'\n",
    "             )\n",
    "print('translated to protein')\n",
    "\n",
    "# run blast\n",
    "blast_cmd_list = ['blastp',\n",
    "                  '-query',\n",
    "                  f'{results_dir}{cell_id}/{cell_id}_protein.fa',\n",
    "                  '-db',\n",
    "                  proteome_ref,\n",
    "                  '-outfmt',\n",
    "                  '\"6 qseqid sseqid evalue\"',\n",
    "                  '-max_target_seqs',\n",
    "                  '1',\n",
    "                  '-out',\n",
    "                  f'{results_dir}{cell_id}/{cell_id}_blastp_out.txt',\n",
    "                  '-num_threads',\n",
    "                  f'{n_threads}'\n",
    "                 ]\n",
    "blast_cmd = ' '.join(blast_cmd_list)\n",
    "print(blast_cmd)\n",
    "subprocess.call(blast_cmd, shell=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>common_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>HG2A_HUMAN</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ATP6_HUMAN</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NU4M_HUMAN</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RS3_HUMAN</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EF1A1_HUMAN</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RS18_HUMAN</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>COX7C_HUMAN</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MGMT_HUMAN</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>DRC6_HUMAN</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SSPN_HUMAN</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RS11_HUMAN</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>FRIH_HUMAN</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TMM68_HUMAN</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>COR1A_HUMAN</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RL13A_HUMAN</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>PDIP3_HUMAN</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>LRC59_HUMAN</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RS14_HUMAN</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>DPB1_HUMAN</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EF2_HUMAN</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RL27A_HUMAN</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>COMD2_HUMAN</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RL21_HUMAN</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CH044_HUMAN</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CH049_HUMAN</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ACTB_HUMAN</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>B2MG_HUMAN</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>EF1A3_HUMAN</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RAB10_HUMAN</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>COX1_HUMAN</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ZNF91_HUMAN</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RLA1_HUMAN</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ACTG_HUMAN</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>HUNIN_HUMAN</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>OXLD1_HUMAN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ATP8_HUMAN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RS9_HUMAN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>UBIM_HUMAN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SPTN1_HUMAN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ZDHC6_HUMAN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IGJ_HUMAN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MTA3_HUMAN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TNFC_HUMAN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NDUAD_HUMAN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RL10_HUMAN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>FA5_HUMAN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NANGN_HUMAN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TYB4_HUMAN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TM243_HUMAN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>PTPM1_HUMAN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             common_name\n",
       "HG2A_HUMAN           170\n",
       "ATP6_HUMAN            88\n",
       "NU4M_HUMAN            68\n",
       "RS3_HUMAN             65\n",
       "EF1A1_HUMAN           60\n",
       "RS18_HUMAN            60\n",
       "COX7C_HUMAN           41\n",
       "MGMT_HUMAN            39\n",
       "DRC6_HUMAN            25\n",
       "SSPN_HUMAN            25\n",
       "RS11_HUMAN            24\n",
       "FRIH_HUMAN            23\n",
       "TMM68_HUMAN           22\n",
       "COR1A_HUMAN           20\n",
       "RL13A_HUMAN           19\n",
       "PDIP3_HUMAN           18\n",
       "LRC59_HUMAN           15\n",
       "RS14_HUMAN            14\n",
       "DPB1_HUMAN            14\n",
       "EF2_HUMAN             12\n",
       "RL27A_HUMAN           12\n",
       "COMD2_HUMAN           11\n",
       "RL21_HUMAN            11\n",
       "CH044_HUMAN           10\n",
       "CH049_HUMAN            9\n",
       "ACTB_HUMAN             9\n",
       "B2MG_HUMAN             8\n",
       "EF1A3_HUMAN            8\n",
       "RAB10_HUMAN            7\n",
       "COX1_HUMAN             7\n",
       "ZNF91_HUMAN            7\n",
       "RLA1_HUMAN             7\n",
       "ACTG_HUMAN             7\n",
       "HUNIN_HUMAN            6\n",
       "OXLD1_HUMAN            5\n",
       "ATP8_HUMAN             5\n",
       "RS9_HUMAN              5\n",
       "UBIM_HUMAN             5\n",
       "SPTN1_HUMAN            5\n",
       "ZDHC6_HUMAN            5\n",
       "IGJ_HUMAN              5\n",
       "MTA3_HUMAN             5\n",
       "TNFC_HUMAN             5\n",
       "NDUAD_HUMAN            5\n",
       "RL10_HUMAN             5\n",
       "FA5_HUMAN              4\n",
       "NANGN_HUMAN            4\n",
       "TYB4_HUMAN             4\n",
       "TM243_HUMAN            4\n",
       "PTPM1_HUMAN            4"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = pd.read_csv('/home/ubuntu/data/longread/blast_results/A10_TSO1/A10_TSO1_blastp_out.txt', \n",
    "                      sep = '\\t',\n",
    "                      header = None\n",
    "                     )\n",
    "results.columns = ['qsid' , 'ssid', 'evalue']\n",
    "results['frame'] = [x.split('__')[-1][-1] for x in results['qsid']]\n",
    "results['qsid'] = [x.split('__')[0] for x in results['qsid']]\n",
    "results['uniprot_id'] = [x.split('|')[1] for x in results['ssid']]\n",
    "results['common_name'] = [x.split('|')[2] for x in results['ssid']]\n",
    "results = results.sort_values('evalue', ascending = True).groupby(['qsid']).head(1)\n",
    "pd.DataFrame(results['common_name'].value_counts()).head(50)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'CD74_HUMAN' in results['common_name'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "322"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(results.qsid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
