
% Default to the notebook output style

    


% Inherit from the specified cell style.




    
\documentclass[11pt]{article}

    
    
    \usepackage[T1]{fontenc}
    % Nicer default font (+ math font) than Computer Modern for most use cases
    \usepackage{mathpazo}

    % Basic figure setup, for now with no caption control since it's done
    % automatically by Pandoc (which extracts ![](path) syntax from Markdown).
    \usepackage{graphicx}
    % We will generate all images so they have a width \maxwidth. This means
    % that they will get their normal width if they fit onto the page, but
    % are scaled down if they would overflow the margins.
    \makeatletter
    \def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth
    \else\Gin@nat@width\fi}
    \makeatother
    \let\Oldincludegraphics\includegraphics
    % Set max figure width to be 80% of text width, for now hardcoded.
    \renewcommand{\includegraphics}[1]{\Oldincludegraphics[width=.8\maxwidth]{#1}}
    % Ensure that by default, figures have no caption (until we provide a
    % proper Figure object with a Caption API and a way to capture that
    % in the conversion process - todo).
    \usepackage{caption}
    \DeclareCaptionLabelFormat{nolabel}{}
    \captionsetup{labelformat=nolabel}

    \usepackage{adjustbox} % Used to constrain images to a maximum size 
    \usepackage{xcolor} % Allow colors to be defined
    \usepackage{enumerate} % Needed for markdown enumerations to work
    \usepackage{geometry} % Used to adjust the document margins
    \usepackage{amsmath} % Equations
    \usepackage{amssymb} % Equations
    \usepackage{textcomp} % defines textquotesingle
    % Hack from http://tex.stackexchange.com/a/47451/13684:
    \AtBeginDocument{%
        \def\PYZsq{\textquotesingle}% Upright quotes in Pygmentized code
    }
    \usepackage{upquote} % Upright quotes for verbatim code
    \usepackage{eurosym} % defines \euro
    \usepackage[mathletters]{ucs} % Extended unicode (utf-8) support
    \usepackage[utf8x]{inputenc} % Allow utf-8 characters in the tex document
    \usepackage{fancyvrb} % verbatim replacement that allows latex
    \usepackage{grffile} % extends the file name processing of package graphics 
                         % to support a larger range 
    % The hyperref package gives us a pdf with properly built
    % internal navigation ('pdf bookmarks' for the table of contents,
    % internal cross-reference links, web links for URLs, etc.)
    \usepackage{hyperref}
    \usepackage{longtable} % longtable support required by pandoc >1.10
    \usepackage{booktabs}  % table support for pandoc > 1.12.2
    \usepackage[inline]{enumitem} % IRkernel/repr support (it uses the enumerate* environment)
    \usepackage[normalem]{ulem} % ulem is needed to support strikethroughs (\sout)
                                % normalem makes italics be italics, not underlines
    \usepackage{mathrsfs}
    

    
    
    % Colors for the hyperref package
    \definecolor{urlcolor}{rgb}{0,.145,.698}
    \definecolor{linkcolor}{rgb}{.71,0.21,0.01}
    \definecolor{citecolor}{rgb}{.12,.54,.11}

    % ANSI colors
    \definecolor{ansi-black}{HTML}{3E424D}
    \definecolor{ansi-black-intense}{HTML}{282C36}
    \definecolor{ansi-red}{HTML}{E75C58}
    \definecolor{ansi-red-intense}{HTML}{B22B31}
    \definecolor{ansi-green}{HTML}{00A250}
    \definecolor{ansi-green-intense}{HTML}{007427}
    \definecolor{ansi-yellow}{HTML}{DDB62B}
    \definecolor{ansi-yellow-intense}{HTML}{B27D12}
    \definecolor{ansi-blue}{HTML}{208FFB}
    \definecolor{ansi-blue-intense}{HTML}{0065CA}
    \definecolor{ansi-magenta}{HTML}{D160C4}
    \definecolor{ansi-magenta-intense}{HTML}{A03196}
    \definecolor{ansi-cyan}{HTML}{60C6C8}
    \definecolor{ansi-cyan-intense}{HTML}{258F8F}
    \definecolor{ansi-white}{HTML}{C5C1B4}
    \definecolor{ansi-white-intense}{HTML}{A1A6B2}
    \definecolor{ansi-default-inverse-fg}{HTML}{FFFFFF}
    \definecolor{ansi-default-inverse-bg}{HTML}{000000}

    % commands and environments needed by pandoc snippets
    % extracted from the output of `pandoc -s`
    \providecommand{\tightlist}{%
      \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
    \DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
    % Add ',fontsize=\small' for more characters per line
    \newenvironment{Shaded}{}{}
    \newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
    \newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.56,0.13,0.00}{{#1}}}
    \newcommand{\DecValTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\FloatTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\CharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\StringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\CommentTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textit{{#1}}}}
    \newcommand{\OtherTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{{#1}}}
    \newcommand{\AlertTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
    \newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.02,0.16,0.49}{{#1}}}
    \newcommand{\RegionMarkerTok}[1]{{#1}}
    \newcommand{\ErrorTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
    \newcommand{\NormalTok}[1]{{#1}}
    
    % Additional commands for more recent versions of Pandoc
    \newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.53,0.00,0.00}{{#1}}}
    \newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.73,0.40,0.53}{{#1}}}
    \newcommand{\ImportTok}[1]{{#1}}
    \newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.73,0.13,0.13}{\textit{{#1}}}}
    \newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\VariableTok}[1]{\textcolor[rgb]{0.10,0.09,0.49}{{#1}}}
    \newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
    \newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.40,0.40,0.40}{{#1}}}
    \newcommand{\BuiltInTok}[1]{{#1}}
    \newcommand{\ExtensionTok}[1]{{#1}}
    \newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.74,0.48,0.00}{{#1}}}
    \newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.49,0.56,0.16}{{#1}}}
    \newcommand{\InformationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\WarningTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    
    
    % Define a nice break command that doesn't care if a line doesn't already
    % exist.
    \def\br{\hspace*{\fill} \\* }
    % Math Jax compatibility definitions
    \def\gt{>}
    \def\lt{<}
    \let\Oldtex\TeX
    \let\Oldlatex\LaTeX
    \renewcommand{\TeX}{\textrm{\Oldtex}}
    \renewcommand{\LaTeX}{\textrm{\Oldlatex}}
    % Document parameters
    % Document title
    \title{DL20181030\_adult\_mel\_analysis}
    
    
    
    
    

    % Pygments definitions
    
\makeatletter
\def\PY@reset{\let\PY@it=\relax \let\PY@bf=\relax%
    \let\PY@ul=\relax \let\PY@tc=\relax%
    \let\PY@bc=\relax \let\PY@ff=\relax}
\def\PY@tok#1{\csname PY@tok@#1\endcsname}
\def\PY@toks#1+{\ifx\relax#1\empty\else%
    \PY@tok{#1}\expandafter\PY@toks\fi}
\def\PY@do#1{\PY@bc{\PY@tc{\PY@ul{%
    \PY@it{\PY@bf{\PY@ff{#1}}}}}}}
\def\PY#1#2{\PY@reset\PY@toks#1+\relax+\PY@do{#2}}

\expandafter\def\csname PY@tok@w\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\expandafter\def\csname PY@tok@c\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.74,0.48,0.00}{##1}}}
\expandafter\def\csname PY@tok@k\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kt\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.69,0.00,0.25}{##1}}}
\expandafter\def\csname PY@tok@o\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@ow\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\expandafter\def\csname PY@tok@nb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@nf\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@nc\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@nn\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@ne\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.82,0.25,0.23}{##1}}}
\expandafter\def\csname PY@tok@nv\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@no\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.53,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@nl\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.63,0.63,0.00}{##1}}}
\expandafter\def\csname PY@tok@ni\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.60,0.60,0.60}{##1}}}
\expandafter\def\csname PY@tok@na\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.49,0.56,0.16}{##1}}}
\expandafter\def\csname PY@tok@nt\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@nd\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\expandafter\def\csname PY@tok@s\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sd\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@si\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.53}{##1}}}
\expandafter\def\csname PY@tok@se\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.13}{##1}}}
\expandafter\def\csname PY@tok@sr\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.53}{##1}}}
\expandafter\def\csname PY@tok@ss\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@sx\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@m\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@gh\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@gu\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@gd\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@gi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\expandafter\def\csname PY@tok@gr\endcsname{\def\PY@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@ge\endcsname{\let\PY@it=\textit}
\expandafter\def\csname PY@tok@gs\endcsname{\let\PY@bf=\textbf}
\expandafter\def\csname PY@tok@gp\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@go\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.53,0.53,0.53}{##1}}}
\expandafter\def\csname PY@tok@gt\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.27,0.87}{##1}}}
\expandafter\def\csname PY@tok@err\endcsname{\def\PY@bc##1{\setlength{\fboxsep}{0pt}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}
\expandafter\def\csname PY@tok@kc\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kd\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kn\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kr\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@bp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@fm\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@vc\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vg\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vm\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@sa\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sc\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@dl\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@s2\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sh\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@s1\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@mb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mf\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mh\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@il\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mo\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@ch\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cm\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cpf\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@c1\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cs\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}

\def\PYZbs{\char`\\}
\def\PYZus{\char`\_}
\def\PYZob{\char`\{}
\def\PYZcb{\char`\}}
\def\PYZca{\char`\^}
\def\PYZam{\char`\&}
\def\PYZlt{\char`\<}
\def\PYZgt{\char`\>}
\def\PYZsh{\char`\#}
\def\PYZpc{\char`\%}
\def\PYZdl{\char`\$}
\def\PYZhy{\char`\-}
\def\PYZsq{\char`\'}
\def\PYZdq{\char`\"}
\def\PYZti{\char`\~}
% for compatibility with earlier versions
\def\PYZat{@}
\def\PYZlb{[}
\def\PYZrb{]}
\makeatother


    % Exact colors from NB
    \definecolor{incolor}{rgb}{0.0, 0.0, 0.5}
    \definecolor{outcolor}{rgb}{0.545, 0.0, 0.0}



    
    % Prevent overflowing lines due to hard-to-break entities
    \sloppy 
    % Setup hyperref package
    \hypersetup{
      breaklinks=true,  % so long urls are correctly broken across lines
      colorlinks=true,
      urlcolor=urlcolor,
      linkcolor=linkcolor,
      citecolor=citecolor,
      }
    % Slightly bigger margins than the latex defaults
    
    \geometry{verbose,tmargin=1in,bmargin=1in,lmargin=1in,rmargin=1in}
    
    

    \begin{document}
    
    
    \maketitle
    
    

    
    \section{Setup}\label{setup}

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor}2}]:} \PY{c+c1}{\PYZsh{} settings}
        \PY{k+kn}{from} \PY{n+nn}{scanpy\PYZus{}helpers} \PY{k}{import} \PY{o}{*}
        \PY{n+nb}{print} \PY{p}{(}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{scanpy version: }\PY{l+s+si}{\PYZob{}\PYZcb{}}\PY{l+s+s1}{\PYZsq{}}\PY{o}{.}\PY{n}{format}\PY{p}{(}\PY{n}{sc}\PY{o}{.}\PY{n}{\PYZus{}\PYZus{}version\PYZus{}\PYZus{}}\PY{p}{)}\PY{p}{)}
        \PY{n}{warnings}\PY{o}{.}\PY{n}{filterwarnings}\PY{p}{(}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{ignore}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)}
        \PY{o}{\PYZpc{}}\PY{k}{matplotlib} inline
        
        \PY{c+c1}{\PYZsh{} read raw data (expression data) to pandas df}
        \PY{n}{pre\PYZus{}adata} \PY{o}{=} \PY{n}{pd}\PY{o}{.}\PY{n}{read\PYZus{}csv}\PY{p}{(}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{/home/ubuntu/data/DL20181011\PYZus{}melanocyte\PYZus{}test\PYZus{}data/primary\PYZus{}mel\PYZus{}rawdata\PYZus{}181011.csv}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)}
        
        \PY{c+c1}{\PYZsh{} read annotation data}
        \PY{n}{anno} \PY{o}{=} \PY{n}{pd}\PY{o}{.}\PY{n}{read\PYZus{}csv}\PY{p}{(}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{/home/ubuntu/data/DL20181011\PYZus{}melanocyte\PYZus{}test\PYZus{}data/primary\PYZus{}mel\PYZus{}metadata\PYZus{}181011.csv}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)}
\end{Verbatim}

    \begin{Verbatim}[commandchars=\\\{\}]
scanpy version: 1.3.2+78.g5034872

    \end{Verbatim}

    \section{Ingest data protocol}\label{ingest-data-protocol}

\begin{itemize}
\tightlist
\item
  Raw data and annotation CSV provided by Ashley Maynard
\item
  Append gene marker expression
\item
  Subsequently, ERCC-associated gene names are removed.
\end{itemize}

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor}3}]:} \PY{n}{raw\PYZus{}adata} \PY{o}{=} \PY{n}{create\PYZus{}adata}\PY{p}{(}\PY{n}{pre\PYZus{}adata}\PY{p}{)}
        \PY{n}{anno\PYZus{}dict} \PY{o}{=} \PY{p}{\PYZob{}}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{age}\PY{l+s+s1}{\PYZsq{}}\PY{p}{:}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{age}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,}
                     \PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{plate}\PY{l+s+s1}{\PYZsq{}}\PY{p}{:}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{plate}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,}
                     \PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{location}\PY{l+s+s1}{\PYZsq{}}\PY{p}{:}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{Anatomical\PYZus{}location}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,}
                     \PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{race}\PY{l+s+s1}{\PYZsq{}}\PY{p}{:}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{Race}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,}
                     \PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{sex}\PY{l+s+s1}{\PYZsq{}}\PY{p}{:}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{Sex}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,}
                     \PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{color}\PY{l+s+s1}{\PYZsq{}}\PY{p}{:}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{sample\PYZus{}color}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,}
                     \PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{patient}\PY{l+s+s1}{\PYZsq{}}\PY{p}{:}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{patient\PYZus{}id}\PY{l+s+s1}{\PYZsq{}}\PY{p}{\PYZcb{}}
        \PY{n}{append\PYZus{}anno}\PY{p}{(}\PY{n}{raw\PYZus{}adata}\PY{p}{,} \PY{n}{anno}\PY{p}{,} \PY{n}{anno\PYZus{}dict}\PY{p}{)}
        \PY{n}{append\PYZus{}markers}\PY{p}{(}\PY{n}{raw\PYZus{}adata}\PY{p}{,} \PY{n}{gene\PYZus{}markers}\PY{o}{=}\PY{p}{[}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{PMEL}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{KRT1}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{KRT5}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{KRT10}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{TYR}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{MITF}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{HRH2}\PY{l+s+s1}{\PYZsq{}}\PY{p}{]}\PY{p}{)}
        \PY{n}{raw\PYZus{}adata} \PY{o}{=} \PY{n}{remove\PYZus{}ercc}\PY{p}{(}\PY{n}{raw\PYZus{}adata}\PY{p}{)} 
\end{Verbatim}

    \begin{Verbatim}[commandchars=\\\{\}]
Ingest raw data{\ldots}
	Cells: 4356, Genes: 26485
Append annotations{\ldots}
	Cells: 4356, Genes: 26485
Append marker gene expresssion{\ldots}
Remove ERCC genes{\ldots}
Filtered genes: 11
	Cells: 4356, Genes: 26474

    \end{Verbatim}

    \section{Cohort analysis}\label{cohort-analysis}

\begin{itemize}
\tightlist
\item
  Determine general statistics
\item
  Determine feature class statistics
\end{itemize}

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor}4}]:} \PY{n}{input\PYZus{}data} \PY{o}{=} \PY{n}{raw\PYZus{}adata}
        
        \PY{c+c1}{\PYZsh{} Cell distribution among patients}
        \PY{n}{plot\PYZus{}data} \PY{o}{=} \PY{n+nb}{list}\PY{p}{(}\PY{n}{input\PYZus{}data}\PY{o}{.}\PY{n}{obs}\PY{o}{.}\PY{n}{groupby}\PY{p}{(}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{patient}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)}\PY{o}{.}\PY{n}{count}\PY{p}{(}\PY{p}{)}\PY{p}{[}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{plate}\PY{l+s+s1}{\PYZsq{}}\PY{p}{]}\PY{p}{)}
        \PY{n}{df} \PY{o}{=} \PY{n}{pd}\PY{o}{.}\PY{n}{DataFrame}\PY{p}{(}\PY{p}{\PYZob{}}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{counts}\PY{l+s+s1}{\PYZsq{}}\PY{p}{:}\PY{n}{plot\PYZus{}data}\PY{p}{\PYZcb{}}\PY{p}{)}
        
        \PY{n}{plotnine}\PY{o}{.}\PY{n}{options}\PY{o}{.}\PY{n}{figure\PYZus{}size} \PY{o}{=} \PY{p}{(}\PY{l+m+mi}{2}\PY{p}{,}\PY{l+m+mi}{2}\PY{p}{)}
        \PY{n+nb}{print}\PY{p}{(}\PY{n}{ggplot}\PY{p}{(}\PY{n}{df}\PY{p}{,} \PY{n}{aes}\PY{p}{(}\PY{n}{x}\PY{o}{=}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{counts}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)}\PY{p}{)}\PY{o}{+}
             \PY{n}{theme\PYZus{}bw}\PY{p}{(}\PY{p}{)} \PY{o}{+}
             \PY{n}{theme}\PY{p}{(}\PY{n}{aspect\PYZus{}ratio}\PY{o}{=}\PY{l+m+mi}{1}\PY{p}{)}\PY{o}{+}
            \PY{n}{geom\PYZus{}histogram}\PY{p}{(}\PY{p}{)}\PY{o}{+}
            \PY{n}{labs}\PY{p}{(}\PY{n}{x}\PY{o}{=}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{cell count}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,} \PY{n}{y}\PY{o}{=}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{frequency}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)}\PY{p}{)}
        
        \PY{c+c1}{\PYZsh{} Patient distribution among classes}
        \PY{n}{count\PYZus{}list} \PY{o}{=} \PY{p}{[}\PY{p}{]}
        \PY{n}{label\PYZus{}list} \PY{o}{=} \PY{p}{[}\PY{p}{]}
        \PY{n}{group\PYZus{}list} \PY{o}{=} \PY{p}{[}\PY{p}{]}
        \PY{k}{for} \PY{n}{feat} \PY{o+ow}{in} \PY{p}{[}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{age}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{color}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{sex}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{race}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{location}\PY{l+s+s1}{\PYZsq{}}\PY{p}{]}\PY{p}{:}
            \PY{n}{count\PYZus{}list} \PY{o}{=} \PY{n}{count\PYZus{}list} \PY{o}{+} \PY{n}{input\PYZus{}data}\PY{o}{.}\PY{n}{obs}\PY{o}{.}\PY{n}{groupby}\PY{p}{(}\PY{n}{by}\PY{o}{=}\PY{n}{feat}\PY{p}{)}\PY{o}{.}\PY{n}{nunique}\PY{p}{(}\PY{p}{)}\PY{p}{[}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{patient}\PY{l+s+s1}{\PYZsq{}}\PY{p}{]}\PY{o}{.}\PY{n}{values}\PY{o}{.}\PY{n}{tolist}\PY{p}{(}\PY{p}{)}
            \PY{n}{labels} \PY{o}{=} \PY{n}{input\PYZus{}data}\PY{o}{.}\PY{n}{obs}\PY{o}{.}\PY{n}{groupby}\PY{p}{(}\PY{n}{by}\PY{o}{=}\PY{n}{feat}\PY{p}{)}\PY{o}{.}\PY{n}{nunique}\PY{p}{(}\PY{p}{)}\PY{o}{.}\PY{n}{index}\PY{o}{.}\PY{n}{tolist}\PY{p}{(}\PY{p}{)}
            \PY{n}{label\PYZus{}list} \PY{o}{=} \PY{n}{label\PYZus{}list} \PY{o}{+} \PY{n}{labels}
            \PY{n}{group\PYZus{}list} \PY{o}{=} \PY{n}{group\PYZus{}list} \PY{o}{+} \PY{p}{[}\PY{n}{feat} \PY{k}{for} \PY{n}{x} \PY{o+ow}{in} \PY{n+nb}{range}\PY{p}{(}\PY{n+nb}{len}\PY{p}{(}\PY{n}{labels}\PY{p}{)}\PY{p}{)}\PY{p}{]}
        \PY{n}{df} \PY{o}{=} \PY{n}{pd}\PY{o}{.}\PY{n}{DataFrame}\PY{p}{(}\PY{p}{\PYZob{}}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{group}\PY{l+s+s1}{\PYZsq{}}\PY{p}{:}\PY{n}{group\PYZus{}list}\PY{p}{,}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{key}\PY{l+s+s1}{\PYZsq{}}\PY{p}{:}\PY{n}{label\PYZus{}list}\PY{p}{,}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{count}\PY{l+s+s1}{\PYZsq{}}\PY{p}{:}\PY{n}{count\PYZus{}list}\PY{p}{\PYZcb{}}\PY{p}{)}
        
        \PY{n}{plotnine}\PY{o}{.}\PY{n}{options}\PY{o}{.}\PY{n}{figure\PYZus{}size} \PY{o}{=} \PY{p}{(}\PY{l+m+mi}{10}\PY{p}{,} \PY{l+m+mi}{10}\PY{p}{)}
        \PY{n+nb}{print}\PY{p}{(}\PY{n}{ggplot}\PY{p}{(}\PY{n}{df}\PY{p}{,} \PY{n}{aes}\PY{p}{(}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{key}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{count}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)}\PY{p}{)}\PY{o}{+}
             \PY{n}{theme\PYZus{}bw}\PY{p}{(}\PY{p}{)} \PY{o}{+}
             \PY{n}{theme}\PY{p}{(}\PY{n}{aspect\PYZus{}ratio}\PY{o}{=}\PY{l+m+mi}{1}\PY{p}{,}
                  \PY{n}{axis\PYZus{}text\PYZus{}x}\PY{o}{=}\PY{n}{element\PYZus{}text}\PY{p}{(}\PY{n}{angle}\PY{o}{=}\PY{l+m+mi}{90}\PY{p}{)}\PY{p}{,}
                  \PY{n}{panel\PYZus{}spacing\PYZus{}x}\PY{o}{=}\PY{l+m+mf}{0.3}\PY{p}{)} \PY{o}{+}
            \PY{n}{geom\PYZus{}bar}\PY{p}{(}\PY{n}{stat}\PY{o}{=}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{identity}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,} \PY{n}{position}\PY{o}{=}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{dodge}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)} \PY{o}{+}
            \PY{n}{facet\PYZus{}wrap}\PY{p}{(}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{\PYZti{}group}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,} \PY{n}{scales}\PY{o}{=}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{free}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,} \PY{n}{nrow} \PY{o}{=} \PY{l+m+mi}{1}\PY{p}{)} \PY{o}{+}
            \PY{n}{labs}\PY{p}{(}\PY{n}{y}\PY{o}{=}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{pateint count}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,} \PY{n}{x}\PY{o}{=}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)}\PY{p}{)}
\end{Verbatim}

    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{output_5_0.png}
    \end{center}
    { \hspace*{\fill} \\}
    
    \begin{Verbatim}[commandchars=\\\{\}]
<ggplot: (-9223363281151987070)>

    \end{Verbatim}

    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{output_5_2.png}
    \end{center}
    { \hspace*{\fill} \\}
    
    \begin{Verbatim}[commandchars=\\\{\}]
<ggplot: (-9223363281153281031)>

    \end{Verbatim}

    \section{All-cells analysis}\label{all-cells-analysis}

\begin{itemize}
\tightlist
\item
  Filter cells/genes: min-reads, min-genes, and dispersion range
\item
  Scale/transform expression values
\item
  PCA dimensionality reduction
\item
  Cluster by diffusion-like random walk
\item
  Louvain community detection
\item
  UMAP plot
\end{itemize}

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor}5}]:} \PY{n}{full\PYZus{}adata} \PY{o}{=} \PY{n}{process\PYZus{}adata}\PY{p}{(}\PY{n}{raw\PYZus{}adata}\PY{p}{)}
        \PY{n}{pca\PYZus{}adata}\PY{p}{(}\PY{n}{full\PYZus{}adata}\PY{p}{)}
        \PY{n}{umap\PYZus{}adata}\PY{p}{(}\PY{n}{full\PYZus{}adata}\PY{p}{)}
\end{Verbatim}

    \begin{Verbatim}[commandchars=\\\{\}]
Process expression data{\ldots}
	Initial values:
	Cells: 4356, Genes: 26474
min counts per cell(min\_counts): 50000
min genes per cell(min\_genes): 500
min expression dispersion(min\_disp): 0.1
min mean expression(min\_mean): 0.001
max mean expression(max\_mean): 1000.0
Filtered cells: 45
Filtered genes: 15764
	Final values:
	Cells: 4311, Genes: 10710
Principle component analysis{\ldots}

    \end{Verbatim}

    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{output_7_1.png}
    \end{center}
    { \hspace*{\fill} \\}
    
    \begin{Verbatim}[commandchars=\\\{\}]
Enter number of principle components to use:
20
principle\_components(num\_pcs): 20
cells/neighborhood(hoods): 30
Uniform manifold approximation and projection{\ldots}
	Scan Louvain detection resolutions
	resolution\_interval(step\_size): 0.05

    \end{Verbatim}

    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{output_7_3.png}
    \end{center}
    { \hspace*{\fill} \\}
    
    \begin{Verbatim}[commandchars=\\\{\}]
<ggplot: (-9223363281152271342)>
Enter Louvain detection resolution to use:
0.8
resolution(res): 0.8
{\ldots} storing 'age' as categorical
{\ldots} storing 'plate' as categorical
{\ldots} storing 'location' as categorical
{\ldots} storing 'race' as categorical
{\ldots} storing 'sex' as categorical
{\ldots} storing 'color' as categorical
{\ldots} storing 'patient' as categorical

    \end{Verbatim}

    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{output_7_5.png}
    \end{center}
    { \hspace*{\fill} \\}
    
    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{output_7_6.png}
    \end{center}
    { \hspace*{\fill} \\}
    
    \section{Classify MEL vs KRT}\label{classify-mel-vs-krt}

\begin{itemize}
\tightlist
\item
  Manual classification based on marker gene expression
\end{itemize}

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor}6}]:} \PY{c+c1}{\PYZsh{} create list of louvain cluster for each cell label}
        \PY{n}{KRT\PYZus{}int} \PY{o}{=} \PY{p}{[}\PY{l+m+mi}{2}\PY{p}{,}\PY{l+m+mi}{8}\PY{p}{,}\PY{l+m+mi}{10}\PY{p}{,}\PY{l+m+mi}{11}\PY{p}{]}
        \PY{n}{MEL\PYZus{}int} \PY{o}{=} \PY{p}{[}\PY{l+m+mi}{0}\PY{p}{,}\PY{l+m+mi}{1}\PY{p}{,}\PY{l+m+mi}{3}\PY{p}{,}\PY{l+m+mi}{4}\PY{p}{,}\PY{l+m+mi}{5}\PY{p}{,}\PY{l+m+mi}{6}\PY{p}{,}\PY{l+m+mi}{7}\PY{p}{,}\PY{l+m+mi}{9}\PY{p}{]}
        \PY{n}{type\PYZus{}dict} \PY{o}{=} \PY{p}{\PYZob{}}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{KRT}\PY{l+s+s1}{\PYZsq{}}\PY{p}{:}\PY{p}{[}\PY{n+nb}{str}\PY{p}{(}\PY{n}{x}\PY{p}{)} \PY{k}{for} \PY{n}{x} \PY{o+ow}{in} \PY{n}{KRT\PYZus{}int}\PY{p}{]}\PY{p}{,}
                     \PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{MEL}\PY{l+s+s1}{\PYZsq{}}\PY{p}{:}\PY{p}{[}\PY{n+nb}{str}\PY{p}{(}\PY{n}{x}\PY{p}{)} \PY{k}{for} \PY{n}{x} \PY{o+ow}{in} \PY{n}{MEL\PYZus{}int}\PY{p}{]}\PY{p}{\PYZcb{}}
        
        \PY{c+c1}{\PYZsh{} DUB\PYZus{}int = [9]}
        \PY{c+c1}{\PYZsh{} type\PYZus{}dict = \PYZob{}\PYZsq{}KRT\PYZsq{}:[str(x) for x in KRT\PYZus{}int],}
        \PY{c+c1}{\PYZsh{}              \PYZsq{}MEL\PYZsq{}:[str(x) for x in MEL\PYZus{}int],}
        \PY{c+c1}{\PYZsh{}              \PYZsq{}DUB\PYZsq{}:[str(x) for x in DUB\PYZus{}int]\PYZcb{}}
            
        \PY{n}{classify\PYZus{}type}\PY{p}{(}\PY{n}{raw\PYZus{}adata}\PY{p}{,} \PY{n}{full\PYZus{}adata}\PY{p}{,} \PY{n}{type\PYZus{}dict}\PY{p}{)}
\end{Verbatim}

    \section{Subset and re-analyze}\label{subset-and-re-analyze}

\begin{itemize}
\tightlist
\item
  Subset filter based on categorical clinical metadata and assigned
  cluster: age, location, race, sex, ratio\_label, color, patient,
  louvain
\end{itemize}

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor}8}]:} \PY{n}{feature\PYZus{}dict} \PY{o}{=} \PY{p}{\PYZob{}}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{age}\PY{l+s+s1}{\PYZsq{}}\PY{p}{:}\PY{p}{[}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{Adult}\PY{l+s+s1}{\PYZsq{}}\PY{p}{]}\PY{p}{,} \PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{type}\PY{l+s+s1}{\PYZsq{}}\PY{p}{:}\PY{p}{[}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{MEL}\PY{l+s+s1}{\PYZsq{}}\PY{p}{]}\PY{p}{\PYZcb{}}
        \PY{n}{adata\PYZus{}subset} \PY{o}{=} \PY{n}{subset\PYZus{}adata}\PY{p}{(}\PY{n}{raw\PYZus{}adata}\PY{p}{,} \PY{n}{feature\PYZus{}dict}\PY{p}{)}
        \PY{n}{adata\PYZus{}subset} \PY{o}{=} \PY{n}{process\PYZus{}adata}\PY{p}{(}\PY{n}{adata\PYZus{}subset}\PY{p}{)}
        \PY{n}{pca\PYZus{}adata}\PY{p}{(}\PY{n}{adata\PYZus{}subset}\PY{p}{)}
        \PY{n}{umap\PYZus{}adata}\PY{p}{(}\PY{n}{adata\PYZus{}subset}\PY{p}{)}
\end{Verbatim}

    \begin{Verbatim}[commandchars=\\\{\}]
Subsetting data{\ldots}
key = age, value = Adult, matched = 2658
key = type, value = MEL, matched = 3273
	Cells: 1873, Genes: 26474
Process expression data{\ldots}
	Initial values:
	Cells: 1873, Genes: 26474
min counts per cell(min\_counts): 50000
min genes per cell(min\_genes): 500
min expression dispersion(min\_disp): 0.1
min mean expression(min\_mean): 0.001
max mean expression(max\_mean): 1000.0
Filtered cells: 0
Filtered genes: 16953
	Final values:
	Cells: 1873, Genes: 9521
Principle component analysis{\ldots}

    \end{Verbatim}

    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{output_11_1.png}
    \end{center}
    { \hspace*{\fill} \\}
    
    \begin{Verbatim}[commandchars=\\\{\}]
Enter number of principle components to use:
25
principle\_components(num\_pcs): 25
cells/neighborhood(hoods): 30
Uniform manifold approximation and projection{\ldots}
	Scan Louvain detection resolutions
	resolution\_interval(step\_size): 0.05

    \end{Verbatim}

    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{output_11_3.png}
    \end{center}
    { \hspace*{\fill} \\}
    
    \begin{Verbatim}[commandchars=\\\{\}]
<ggplot: (-9223363281155557260)>
Enter Louvain detection resolution to use:
0.6
resolution(res): 0.6
{\ldots} storing 'age' as categorical
{\ldots} storing 'plate' as categorical
{\ldots} storing 'location' as categorical
{\ldots} storing 'race' as categorical
{\ldots} storing 'sex' as categorical
{\ldots} storing 'color' as categorical
{\ldots} storing 'patient' as categorical
{\ldots} storing 'type' as categorical

    \end{Verbatim}

    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{output_11_5.png}
    \end{center}
    { \hspace*{\fill} \\}
    
    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{output_11_6.png}
    \end{center}
    { \hspace*{\fill} \\}
    
    \section{Quantify class covariance}\label{quantify-class-covariance}

\begin{itemize}
\tightlist
\item
  High degree of covariation indicates dependent classes
\end{itemize}

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor}9}]:} \PY{c+c1}{\PYZsh{} correlation among features}
        \PY{n}{input\PYZus{}data} \PY{o}{=} \PY{n}{adata\PYZus{}subset}
        
        \PY{n}{fltrd\PYZus{}labels} \PY{o}{=} \PY{p}{[}\PY{n}{key} \PY{k}{for} \PY{n}{key}\PY{p}{,}\PY{n}{value} \PY{o+ow}{in} \PY{n}{anno\PYZus{}dict}\PY{o}{.}\PY{n}{items}\PY{p}{(}\PY{p}{)}\PY{p}{]} \PY{o}{+} \PY{p}{[}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{louvain}\PY{l+s+s1}{\PYZsq{}}\PY{p}{]}
        \PY{n}{df} \PY{o}{=} \PY{n}{pd}\PY{o}{.}\PY{n}{DataFrame}\PY{p}{(}\PY{p}{)}   
        \PY{k}{for} \PY{n}{label\PYZus{}col} \PY{o+ow}{in} \PY{n}{fltrd\PYZus{}labels}\PY{p}{:}
            \PY{n}{acc\PYZus{}list} \PY{o}{=} \PY{p}{[}\PY{p}{]}
            \PY{k}{for} \PY{n}{label\PYZus{}row} \PY{o+ow}{in} \PY{n}{fltrd\PYZus{}labels}\PY{p}{:}
                \PY{n}{acc\PYZus{}list}\PY{o}{.}\PY{n}{append}\PY{p}{(}\PY{n}{class2class\PYZus{}reg}\PY{p}{(}\PY{n}{X}\PY{o}{=}\PY{n}{input\PYZus{}data}\PY{o}{.}\PY{n}{obs}\PY{p}{[}\PY{n}{label\PYZus{}row}\PY{p}{]}\PY{p}{,} \PY{n}{y}\PY{o}{=}\PY{n}{input\PYZus{}data}\PY{o}{.}\PY{n}{obs}\PY{p}{[}\PY{n}{label\PYZus{}col}\PY{p}{]}\PY{p}{)}\PY{p}{)}
            \PY{n}{df}\PY{p}{[}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+si}{\PYZob{}\PYZcb{}}\PY{l+s+s1}{\PYZsq{}}\PY{o}{.}\PY{n}{format}\PY{p}{(}\PY{n}{label\PYZus{}col}\PY{p}{)}\PY{p}{]} \PY{o}{=} \PY{n}{acc\PYZus{}list}
        \PY{n}{df}\PY{p}{[}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{predictor}\PY{l+s+s1}{\PYZsq{}}\PY{p}{]} \PY{o}{=} \PY{n}{fltrd\PYZus{}labels}
        \PY{c+c1}{\PYZsh{} print(df)}
        
        \PY{n}{plotnine}\PY{o}{.}\PY{n}{options}\PY{o}{.}\PY{n}{figure\PYZus{}size} \PY{o}{=} \PY{p}{(}\PY{l+m+mi}{3}\PY{p}{,}\PY{l+m+mi}{3}\PY{p}{)}
        \PY{n}{plot\PYZus{}df} \PY{o}{=} \PY{n}{pd}\PY{o}{.}\PY{n}{melt}\PY{p}{(}\PY{n}{df}\PY{p}{,} \PY{n}{id\PYZus{}vars}\PY{o}{=}\PY{p}{[}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{predictor}\PY{l+s+s1}{\PYZsq{}}\PY{p}{]}\PY{p}{)}
        \PY{p}{(}\PY{n}{ggplot}\PY{p}{(}\PY{n}{plot\PYZus{}df}\PY{p}{,} \PY{n}{aes}\PY{p}{(}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{variable}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{predictor}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,}\PY{n}{fill}\PY{o}{=}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{value}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)}\PY{p}{)} \PY{o}{+}
             \PY{n}{theme\PYZus{}bw}\PY{p}{(}\PY{p}{)} \PY{o}{+}
             \PY{n}{theme}\PY{p}{(}\PY{n}{aspect\PYZus{}ratio}\PY{o}{=}\PY{l+m+mi}{1}\PY{p}{,}
                  \PY{n}{axis\PYZus{}text\PYZus{}x}\PY{o}{=}\PY{n}{element\PYZus{}text}\PY{p}{(}\PY{n}{angle}\PY{o}{=}\PY{l+m+mi}{90}\PY{p}{)}\PY{p}{)} \PY{o}{+}
             \PY{n}{geom\PYZus{}tile}\PY{p}{(}\PY{p}{)} \PY{o}{+}
             \PY{n}{labs}\PY{p}{(}\PY{n}{x}\PY{o}{=}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{response}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,} \PY{n}{y}\PY{o}{=}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{predictor}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,} \PY{n}{fill}\PY{o}{=}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{F1\PYZhy{}score}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)}\PY{o}{+}
            \PY{n}{scale\PYZus{}fill\PYZus{}cmap}\PY{p}{(}\PY{n}{name}\PY{o}{=}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{afmhot}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)}\PY{p}{)}
\end{Verbatim}

    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{output_13_0.png}
    \end{center}
    { \hspace*{\fill} \\}
    
\begin{Verbatim}[commandchars=\\\{\}]
{\color{outcolor}Out[{\color{outcolor}9}]:} <ggplot: (8755698469524)>
\end{Verbatim}
            
    \section{Differential gene expression rank per
cluster}\label{differential-gene-expression-rank-per-cluster}

\begin{itemize}
\tightlist
\item
  Top 50 enriched genes per cluster
\end{itemize}

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor}157}]:} \PY{n}{input\PYZus{}adata} \PY{o}{=} \PY{n}{adata\PYZus{}subset}
          \PY{n}{sc}\PY{o}{.}\PY{n}{tl}\PY{o}{.}\PY{n}{rank\PYZus{}genes\PYZus{}groups}\PY{p}{(}\PY{n}{input\PYZus{}adata}\PY{p}{,} \PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{louvain}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,} \PY{n}{method}\PY{o}{=}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{logreg}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)}
          \PY{n}{sc}\PY{o}{.}\PY{n}{pl}\PY{o}{.}\PY{n}{rank\PYZus{}genes\PYZus{}groups}\PY{p}{(}\PY{n}{input\PYZus{}adata}\PY{p}{,} \PY{n}{n\PYZus{}genes}\PY{o}{=}\PY{l+m+mi}{50}\PY{p}{,} \PY{n}{sharey}\PY{o}{=}\PY{k+kc}{False}\PY{p}{,} \PY{n}{n\PYZus{}panels\PYZus{}per\PYZus{}row}\PY{o}{=}\PY{l+m+mi}{1}\PY{p}{,} \PY{n}{fontsize}\PY{o}{=}\PY{l+m+mi}{7}\PY{p}{)}
          \PY{n}{pd}\PY{o}{.}\PY{n}{DataFrame}\PY{p}{(}\PY{n}{input\PYZus{}adata}\PY{o}{.}\PY{n}{uns}\PY{p}{[}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{rank\PYZus{}genes\PYZus{}groups}\PY{l+s+s1}{\PYZsq{}}\PY{p}{]}\PY{p}{[}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{names}\PY{l+s+s1}{\PYZsq{}}\PY{p}{]}\PY{p}{)}\PY{o}{.}\PY{n}{head}\PY{p}{(}\PY{l+m+mi}{50}\PY{p}{)}
\end{Verbatim}

    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{output_15_0.png}
    \end{center}
    { \hspace*{\fill} \\}
    
\begin{Verbatim}[commandchars=\\\{\}]
{\color{outcolor}Out[{\color{outcolor}157}]:}                0             1             2               3             4  \textbackslash{}
          0           HRH2          MT1X          GZMA          ZNF219  LOC100129046   
          1           COIL       RAPGEF3          NKG7          ADRA2C          BHMT   
          2        C3orf79        DDIT4L        CORO1A           KIF17       SNORD53   
          3         MESDC2          CD63          RAC2           CDIP1      CRISPLD1   
          4          CSPG5        GPR132        CARD16          SPTAN1      GOLGA2P5   
          5         EXOSC3       S100A10         SKAP1          ZBTB46       MIR4526   
          6          PTGDS         TCEA3          RGS1          SLC7A4         NLRC4   
          7         MRPL15          HHIP          CD3D          RBM15B       SLC37A2   
          8          CPXM2        LGALS1           HPN           OR7D2  LOC101929441   
          9         ZNF749        S100A6        IL10RA         UXT-AS1          SAT2   
          10        PNPLA4          NPC2           CD2     EFCAB14-AS1          DBX2   
          11        COMMD6        LGALS3       ALOX5AP           ATHL1     LINC00618   
          12         RTN4R      SLC22A11          SIT1       CLYBL-AS1        SEMA4G   
          13          NFU1         SUSD4          CD8A           SCIMP        MAATS1   
          14        CCSER2         IL23A      ITPRIPL1           KCNJ9        HTT-AS   
          15      TSGA10IP         S1PR3        AMICA1          ITGBL1        ASPRV1   
          16          SNCB       WFIKKN1         GPR65          KCNIP4         MYO1F   
          17         IFI27     LOC442132         PRDM1         MIR548N        SPATA4   
          18         PAICS        GRIN3A         CYTIP            EHD4          CD72   
          19          PKIG        COL5A2      C16orf54            TBK1        OR52I1   
          20        TRIM44         CLCA4          CD69         GTF2IP1         NMUR1   
          21         FXYD3         TBX15        RASSF5          PRSS56          HID1   
          22          ANK2       TSPAN11       APBB1IP    LOC101927066         NRIP2   
          23         SUGT1     LINC01132          CD3G    LOC101060542         HSPA6   
          24         PRR26  LOC101927188           LCK    LOC101929470  LOC101928030   
          25        LGALS7      FLJ43315         IGLL5           DHX35       TUBA3FP   
          26         CHST9     LINC00310        ZNF683          HIATL2         TREX2   
          27          ST20          CD19         RCAN2          DFNB31       MIR4794   
          28  LOC100288152          BMP6        PTPN22           NR4A3          LGR6   
          29         ITGAE        MGAT5B           LTB    LOC101928896         SAXO1   
          30         TMOD3         SRP14        CCL4L1       TNFRSF10A   SYS1-DBNDD2   
          31         LAMC1      HHIP-AS1      TBC1D10C           BEST4  LOC102724084   
          32        ZNF511        COL3A1          GZMM           ZNF20          GNAL   
          33        NSMCE1       TMEM63C        RAB27B    LOC101929633        PODNL1   
          34          HUS1       WDR83OS           FYB          ZBTB22      C17orf96   
          35       NDUFAF5          KPRP        ZNF625          SCART1          APOD   
          36    ERICH6-AS1          RPS8       SNORA75         SHROOM1  NAALADL2-AS1   
          37         CCPG1          TTPA        LY6G6C    ARHGEF26-AS1     LINC00540   
          38      FAM133CP      IL1RAPL2         PGBD2        ZEB1-AS1         VAMP5   
          39     RPL13AP20        TSPAN2       SLC38A1            CLP1         NOC2L   
          40       AFG3L1P         GPM6B         DUSP9           HOXA1        USP27X   
          41         YIF1B         CBLN4      ANKRD18A  TCONS\_00029157        HNRNPC   
          42         HAUS8         PEBP1         LIME1         SLC22A5        MIR593   
          43         TRPM4         PTPN7        ZBTB8A         MIR4716       MIR3934   
          44         CRYAB          SSX8       ARHGDIB        PRC1-AS1      MIR4666B   
          45        KDELR3         IQCA1          KRT1          HIGD1B      HIST1H4A   
          46     LINC01443     LINC00907  LOC101927865      RHOXF1-AS1         OTOL1   
          47          ERN1         STMN1       IL12RB1         SCARNA4       MIR6758   
          48         POLE4        ARMC12         FANCI    LOC101928270           PYY   
          49        TBC1D7        TXNL4B        MFSD2B           ZFP64         ENPP7   
          
                        5  
          0       COL17A1  
          1           FST  
          2         KRT15  
          3        IL20RA  
          4         LAMB4  
          5       GYLTL1B  
          6         KRT31  
          7       OR4F13P  
          8         LOXL4  
          9         MROH7  
          10        TAF7L  
          11        MXRA5  
          12         FMOD  
          13         EDN2  
          14         NRG1  
          15         ORC1  
          16        OR4X1  
          17    LOC442497  
          18      SLCO1C1  
          19      MIR8059  
          20         MDFI  
          21         XIST  
          22      COL14A1  
          23        INHBB  
          24  MACROD2-AS1  
          25      MIR4279  
          26        POSTN  
          27      LDHAL6B  
          28          CPE  
          29        LCE2B  
          30         MEST  
          31       FERMT1  
          32         DLK2  
          33    LOC146880  
          34      MIR4431  
          35         NEU2  
          36       S100A2  
          37     MIR205HG  
          38       CXCL14  
          39         NPR3  
          40    PARD3-AS1  
          41        KRT17  
          42        VSIG8  
          43      SNORA72  
          44      ANXA2P1  
          45         GJB6  
          46         RBP5  
          47      C1orf64  
          48    C17orf105  
          49        TRPC6  
\end{Verbatim}
            
    \section{Candidate gene contribution to principle
components}\label{candidate-gene-contribution-to-principle-components}

\begin{itemize}
\tightlist
\item
  PCA loadings are normalized to max PC loading
\end{itemize}

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor}158}]:} \PY{n}{target} \PY{o}{=} \PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{HRH2}\PY{l+s+s1}{\PYZsq{}}
          \PY{n}{loadings} \PY{o}{=} \PY{n}{adata\PYZus{}subset}\PY{p}{[}\PY{p}{:}\PY{p}{,}\PY{n}{target}\PY{p}{]}\PY{o}{.}\PY{n}{varm}\PY{p}{[}\PY{l+m+mi}{0}\PY{p}{]}\PY{p}{[}\PY{l+m+mi}{0}\PY{p}{]}\PY{o}{.}\PY{n}{tolist}\PY{p}{(}\PY{p}{)}
          \PY{n}{max\PYZus{}loadings} \PY{o}{=} \PY{p}{[}\PY{n}{np}\PY{o}{.}\PY{n}{max}\PY{p}{(}\PY{n}{np}\PY{o}{.}\PY{n}{array}\PY{p}{(}\PY{p}{[}\PY{n}{adata\PYZus{}subset}\PY{o}{.}\PY{n}{varm}\PY{p}{[}\PY{n}{row}\PY{p}{]}\PY{p}{[}\PY{l+m+mi}{0}\PY{p}{]}\PY{p}{[}\PY{n}{pc}\PY{p}{]} \PY{k}{for} \PY{n}{row} \PY{o+ow}{in} \PY{n+nb}{range}\PY{p}{(}\PY{n}{adata\PYZus{}subset}\PY{o}{.}\PY{n}{varm}\PY{o}{.}\PY{n}{shape}\PY{p}{[}\PY{l+m+mi}{0}\PY{p}{]}\PY{p}{)}\PY{p}{]}\PY{p}{)}\PY{p}{)} \PY{k}{for} \PY{n}{pc} \PY{o+ow}{in} \PY{n+nb}{range}\PY{p}{(}\PY{l+m+mi}{50}\PY{p}{)}\PY{p}{]}
          \PY{n}{norm\PYZus{}loadings} \PY{o}{=} \PY{p}{[}\PY{n}{x}\PY{o}{/}\PY{n}{y} \PY{k}{for} \PY{n}{x}\PY{p}{,}\PY{n}{y} \PY{o+ow}{in} \PY{n+nb}{zip}\PY{p}{(}\PY{n}{loadings}\PY{p}{,} \PY{n}{max\PYZus{}loadings}\PY{p}{)}\PY{p}{]}
          \PY{n}{plot\PYZus{}df} \PY{o}{=} \PY{n}{pd}\PY{o}{.}\PY{n}{DataFrame}\PY{p}{(}\PY{p}{\PYZob{}}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{norm\PYZus{}loadings}\PY{l+s+s1}{\PYZsq{}}\PY{p}{:}\PY{n}{norm\PYZus{}loadings}\PY{p}{,} 
                                 \PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{PCs}\PY{l+s+s1}{\PYZsq{}}\PY{p}{:}\PY{p}{[}\PY{n}{x} \PY{k}{for} \PY{n}{x} \PY{o+ow}{in} \PY{n+nb}{range}\PY{p}{(}\PY{l+m+mi}{50}\PY{p}{)}\PY{p}{]}\PY{p}{\PYZcb{}}\PY{p}{)}
          
          \PY{n}{plotnine}\PY{o}{.}\PY{n}{options}\PY{o}{.}\PY{n}{figure\PYZus{}size} \PY{o}{=} \PY{p}{(}\PY{l+m+mi}{5}\PY{p}{,}\PY{l+m+mi}{5}\PY{p}{)}
          \PY{p}{(}\PY{n}{ggplot}\PY{p}{(}\PY{n}{plot\PYZus{}df}\PY{p}{,} \PY{n}{aes}\PY{p}{(}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{PCs}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,} \PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{norm\PYZus{}loadings}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)}\PY{p}{)}\PY{o}{+}
               \PY{n}{theme\PYZus{}bw}\PY{p}{(}\PY{p}{)} \PY{o}{+}
               \PY{n}{theme}\PY{p}{(}\PY{n}{aspect\PYZus{}ratio} \PY{o}{=} \PY{l+m+mi}{1}\PY{p}{)} \PY{o}{+}
               \PY{n}{geom\PYZus{}bar}\PY{p}{(}\PY{n}{stat}\PY{o}{=}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{identity}\PY{l+s+s1}{\PYZsq{}}\PY{p}{,} \PY{n}{position}\PY{o}{=}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{dodge}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)} \PY{o}{+}
               \PY{n}{labs}\PY{p}{(}\PY{n}{y}\PY{o}{=}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{relative fractional loading}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)}\PY{p}{)}
\end{Verbatim}

    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{output_17_0.png}
    \end{center}
    { \hspace*{\fill} \\}
    
\begin{Verbatim}[commandchars=\\\{\}]
{\color{outcolor}Out[{\color{outcolor}158}]:} <ggplot: (-9223363281152864487)>
\end{Verbatim}
            

    % Add a bibliography block to the postdoc
    
    
    
    \end{document}
